internal/controllers/s3userclaim/provisioner.go

```diff
@@ -311,3 +342,73 @@ func (r *Reconciler) assembleCephUserSecret(userName, secretName string) (*corev
 
 	return secret, nil
 }
+
+func (r *Reconciler) generateSubUserAccessMap(desiredSubUsers []string,
+	currentSubUsers []admin.SubuserSpec) map[string]string {
+	// Create a hashmap to move all spec and ceph subUsers to it
+	subUserFullIdAccessMap := make(map[string]string)
+
+	// Tag specSubUsers with "create"
+	for _, subUser := range desiredSubUsers {
+		cephSubUserFullId := generateSubUserFullId(r.cephUserFullId, subUser)
+		subUserFullIdAccessMap[cephSubUserFullId] = consts.SubUserTagCreate
+	}
+
+	// Add read-only subUser to subUsers to prevent if from removing
+	subUserFullIdAccessMap[r.readonlyCephUserFullId] = consts.SubUserTagCreate
+
+	// Tag cephSubUsers as remove if they are not already in the hashmap and remove them otherwise
+	// since they are already available on ceph and not needed to created.
+	for _, currentSubUser := range r.cephUser.Subusers {
+		_, exists := subUserFullIdAccessMap[currentSubUser.Name]
+		if exists {
+			delete(subUserFullIdAccessMap, currentSubUser.Name)
+		} else {
+			subUserFullIdAccessMap[currentSubUser.Name] = consts.SubUserTagRemove
+		}
+	}
+	return subUserFullIdAccessMap
+}
+
+func (r *Reconciler) generateSubUser(ctx context.Context, cephUserFullId string,
+	desiredSubuser admin.SubuserSpec) error {
+	r.logger.Info(fmt.Sprintf("Create subUser: %s", desiredSubuser.Name))
+	if err := r.rgwClient.CreateSubuser(ctx, admin.User{ID: cephUserFullId}, desiredSubuser); err != nil {
+		r.logger.Error(err, "failed to create subUser")
+		return err
+	}
+	return nil
+}
+
+func (r *Reconciler) RemoveSubuserAndSecret(ctx context.Context, cephUserFullId string,
+	subuserToRemove admin.SubuserSpec) error {
+	err := r.rgwClient.RemoveSubuser(ctx, admin.User{ID: cephUserFullId}, subuserToRemove)
+	r.logger.Info(fmt.Sprintf("Remove subUser: %s", subuserToRemove.Name))
+	if err != nil {
+		r.logger.Error(err, "failed to remove subUser")
+		return err
+	}
```

Comment:
If the log is required, we should move it before calling `RemoveSubuser`.
There might be a reason behind this, though. Is there?

Reply:
No, actually it should be in the past tense: Removed subuser ....
Do you agree?

Reply:
The function name is fine. Thanks.
I mean swapping line 386 with 385 ( and then handling the error inline. i.e. `if err := ...; err != nil {}`).

Reply:
Correct. Done